src/software/amazon/ion/impl/lite/IonSequenceLite.java

```diff
@@ -398,4 +400,330 @@ void writeBodyTo(IonWriter writer, SymbolTableProvider symbolTableProvider)
             writer.stepOut();
         }
     }
+
+    private class SubListView implements List<IonValue> {
+        private final List<IonValue> parent;
+        private final int fromIndex;
+        private int size;
+        private int structuralModificationCount;
+
+        private SubListView(final List<IonValue> parent,
+                            final int fromIndex,
+                            final int toIndex) {
+            this.parent = parent;
+            this.fromIndex = fromIndex;
+            this.size = toIndex - fromIndex;
+            this.structuralModificationCount = IonSequenceLite.this.structuralModificationCount;
+        }
+
+        public int size() {
+            checkForParentModification();
+            return size;
+        }
+
+        public boolean isEmpty() {
+            checkForParentModification();
+
+            return size == 0;
+        }
+
+        public IonValue get(final int index) {
+            rangeCheck(index);
+            checkForParentModification();
+
+            return parent.get(toParentIndex(index));
+        }
+
+        public IonValue set(final int index, final IonValue element) {
+            rangeCheck(index);
+            checkForParentModification();
+
+            return parent.set(toParentIndex(index), element);
+        }
+
+        public boolean contains(final Object o) {
+            checkForParentModification();
+            return indexOf(o) != -1;
+        }
+
+        public boolean containsAll(final Collection<?> c) {
+            for (Object o : c) {
+                if(indexOf(o) < 0) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public IonValue[] toArray() {
+            checkForParentModification();
+
+            if (size < 1) return EMPTY_VALUE_ARRAY;
+
+            IonValue[] array = new IonValue[size];
+
+            if(parent instanceof IonSequenceLite) {
```

Comment:
The `parent` can never be another type of IonContainer apart from `IonSequenceLite` 

This check is to separate between direct sublists of an `IonSequence` and nested sublists,  e.g.: `sequence.sublist(1,10).sublist(3,9);`. In this case calculating the top level `IonSequenceLite` indexes is not trivial as we need to do it recursively  

Thinking a bit more about it might be good to actually calculate `fromIndex` based on the top level parent so nested sublists can't directly call the parent instead of going through multiple levels of method calls. Will make this change and then we'll be able to use `arraycopy` always

Reply:
nice!